fsck/fsck.c

```diff
@@ -103,21 +103,32 @@ static void usage(char *name)
 })
 
 static int check_clus_chain(struct exfat_de_iter *de_iter,
-			    struct exfat_inode *node)
+				struct exfat_inode *node)
 {
 	struct exfat *exfat = de_iter->exfat;
 	struct exfat_dentry *stream_de;
-	clus_t clus, prev, next;
+	clus_t clus, prev, next, new_clus;
 	uint64_t count, max_count;
+	int err;
 
 	clus = node->first_clus;
 	prev = EXFAT_EOF_CLUSTER;
 	count = 0;
 	max_count = DIV_ROUND_UP(node->size, exfat->clus_size);
 
-	if (node->size == 0 && node->first_clus == EXFAT_FREE_CLUSTER)
-		return 0;
-
+	if (node->size == 0 && node->first_clus == EXFAT_FREE_CLUSTER) {
+		/* locate a cluster for the empty dir if the dir starts with EXFAT_FREE_CLUSTER */
+		if (node->attr & ATTR_SUBDIR) {
+			if (repair_file_ask(de_iter, node,
+					ER_DE_FIRST_CLUS,
+					"size %#" PRIx64 ", but the first cluster %#x",
+					node->size, node->first_clus))
+				goto allocate_cluster;
```

Comment:
I'm not sure why you use goto statement instead of creating static function and call. There is no reason why we should use goto here.

Reply:
In fact, I'm not sure why we should use goto truncate_file either. I just guess that we are suppose to handle the repair process in this way.

Reply:
There is a few truncate_file goto statement(i.e. not only one) in this function. It could be factored out to helper function. but It is not bad even if goto statement is still used.

Reply:
Got it. Thanks for the explanation! I think maybe we can use goto this time?

Reply:
hm... okay...